kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,700 @@
+			c.ipPools[kvp.Key.String()] = *pool
+		}
+	} else {
+		delete(c.ipPools, kvp.Key.String())
+	}
+}
+
+// syncIpam has two main uses. It functions as a garbage collection for leaked IP addresses from Service LoadBalancer
+// The other use case is to update IPs for any Service LoadBalancer that do not have IPs assigned, this could be caused by the user
+// creating Service LoadBalancer before any valid pools were created
+func (c *loadBalancerController) syncIPAM() {
+	// Garbage collection
+	// Check all ipamBlocks with loadBalancer affinity
+	for _, block := range c.ipamBlocks {
+		attributes := block.Value.(*model.AllocationBlock).Attributes
+		for _, attr := range attributes {
+			obj, empty := c.lbResourceCache.Get(*attr.AttrPrimary)
+			service := obj.(*v1.Service)
+			if !empty {
+				// Service with handle exists, we need to check that all assigned IPs with the handle are still in use by the service
+				log.Infof("Service found for handle: %s. Check if all IPs allocated by the handle are in use.", *attr.AttrPrimary)
+				ips, err := c.calicoClient.IPAM().IPsByHandle(c.ctx, *attr.AttrPrimary)
+				if err != nil {
+					log.Errorf("Error getting IPs for handle: %s", *attr.AttrPrimary)
+					break
+				}
+				for _, ingressIP := range service.Status.LoadBalancer.Ingress {
+					inUse := false
+					for _, handleIP := range ips {
+						if handleIP.String() == ingressIP.IP {
+							inUse = true
+						}
+						if !inUse {
+							releaseOptions := ipam.ReleaseOptions{
+								Address: ingressIP.IP,
+							}
+							_, err = c.calicoClient.IPAM().ReleaseIPs(c.ctx, releaseOptions)
+							if err != nil {
+								log.Errorf("Error releasing IP(%s) for service: %s", ingressIP.IP, service.Name)
+							}
+						}
+					}
+				}
+			} else {
+				// Service no longer exists, leak confirmed. Release all IPs allocated with the specific handle
+				log.Infof("Service not found for handle: %s. Releasing unused IPs", *attr.AttrPrimary)
+				err := c.releaseIP(*attr.AttrPrimary)
+				if err != nil {
+					log.Errorf("Error releasing IPAM for handle %s: %s", *attr.AttrPrimary, err)
+				}
+			}
+		}
+	}
+
+	// Check that all services have assigned IPs as requested, skip if there are no ippools
+	if len(c.ipPools) != 0 {
+		for _, key := range c.lbResourceCache.ListKeys() {
+			obj, empty := c.lbResourceCache.Get(key)
+			if !empty {
+				service := obj.(*v1.Service)
+				if isCalicoManagedLoadBalancer(service, c.cfg.AssignIPs) {
+					if service.Status.LoadBalancer.Ingress == nil ||
+						(len(service.Status.LoadBalancer.Ingress) == 1 &&
+							(*service.Spec.IPFamilyPolicy == v1.IPFamilyPolicyRequireDualStack) || *service.Spec.IPFamilyPolicy == v1.IPFamilyPolicyPreferDualStack) {
+						err := c.assignIP(service)
+						if err != nil {
+							log.Errorf("Error assigning IP to service %s: %s", service.Name, err)
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func (c *loadBalancerController) runWorker() {
```

Comment:
I think we can get rid of this entire call chain - all of the Service IP allocation should be happening within `syncIPAM`

It's imperative that we use a single goroutine for both allocating service IPs, as well as IPAM garbage collection, so that the two are always in sync with each other.